zbus: C++ Support

While trying to work on some C++ code that uses ZBUS, I noted that the code would give undefined reference errors when compiling to code for my target platform. When converting the file back to C, these errors dissapear and it works perfectly fine. The code was not using any C++ specific functions making the change from C++ to C very simple (changing the file extension). I was hoping to start using the C++ language features later in in the development.

Key details:
- Board: esp32s3_devkitm

prj.conf:
```
CONFIG_ASSERT=y

CONFIG_CPP=y
CONFIG_GLIBCXX_LIBCPP=y

# ZBUS Configs
CONFIG_ZBUS=y
CONFIG_ZBUS_CHANNEL_NAME=y
CONFIG_ZBUS_OBSERVER_NAME=y
CONFIG_ZBUS_MSG_SUBSCRIBER=y
CONFIG_ZBUS_RUNTIME_OBSERVERS=y
CONFIG_ZBUS_LOG_LEVEL_DBG=n
CONFIG_ZBUS_LOG_LEVEL_INF=y

# Logs
CONFIG_LOG=y

CONFIG_SHELL=y

CONFIG_HEAP_MEM_POOL_SIZE=2048
```
CMakeLists.txt:
```
cmake_minimum_required(VERSION 3.20.0)

find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
project(zbus_tutorial)

FILE(GLOB app_sources src/*.cpp)
target_sources(app PRIVATE ${app_sources})
target_include_directories(app PRIVATE inc)
```
main.cpp:
```cpp

#include <zephyr/init.h>
#include <zephyr/kernel.h>
#include <zephyr/zbus/zbus.h>
LOG_MODULE_DECLARE (zbus, CONFIG_ZBUS_LOG_LEVEL);

ZBUS_CHAN_DECLARE (sensor_data_chan, chan_wifi_connection);

// TODO add thermal skel thread
ZBUS_OBS_DECLARE (sensor_sampler_thread_sub, dev_log_lis);

ZBUS_CHAN_DEFINE (
    sample_trigger_chan, bool,
    NULL, // TODO add validator because we expect to use more triggers
    NULL, ZBUS_OBSERVERS (sensor_sampler_thread_sub), false);

ZBUS_CHAN_ADD_OBS (sensor_data_chan, dev_log_lis, 3);
ZBUS_CHAN_ADD_OBS (chan_wifi_connection, dev_log_lis, 3);

// Assume main will trigger a timer to notify a sample_trigger_chan
int main (void){
    ....
}
```
wifi_thread.cpp:
```cpp
#include <zephyr/kernel.h>
#include <zephyr/kernel/thread.h>
#include <zephyr/zbus/zbus.h>
#include "wifi_thread.h"

// ZBus definitions
ZBUS_CHAN_DEFINE (chan_wifi_connection, bool, NULL, NULL, ZBUS_OBSERVERS_EMPTY, ZBUS_MSG_INIT (false));
ZBUS_OBS_DECLARE (sensor_sampler_thread_sub);
ZBUS_CHAN_ADD_OBS (chan_wifi_connection, sensor_sampler_thread_sub, 3);

// Assume thread runs and on WiFi connect the following line runs:
int thread_main(){
    ...
    wifi_connected_status = true;
    zbus_chan_pub (&chan_wifi_connection, &wifi_connected_status, K_MSEC (500));
    ....
}
```
temp_ctrl_thread.cpp:
```cpp
#include <zephyr/init.h>
#include <zephyr/kernel.h>
#include <zephyr/logging/log.h>
#include <zephyr/zbus/zbus.h>
#include "messages.h"

ZBUS_CHAN_DECLARE (sensor_data_chan);
ZBUS_SUBSCRIBER_DEFINE (temp_ctrl_thread_sub, 4);
ZBUS_CHAN_ADD_OBS (sensor_data_chan, temp_ctrl_thread_sub, 3);

int thread_main(){
  const struct zbus_channel *chan;
  struct sensor_data_msg sdata = { .wetness = 0, .temperature = 0 };

  while (!zbus_sub_wait (&temp_ctrl_thread_sub, &chan, K_FOREVER)){
      if (chan == &sensor_data_chan){
          int err = zbus_chan_read (&sensor_data_chan, &sdata, K_MSEC (500));
          if (err){
              LOG_WRN ("Could not read the channel. Error code: %d", err);
            }
          else{
               LOG_INF ("TEMP_CTRL_THREAD: %u, %u \n", sdata.wetness , sdata.temperature);
            }
        }
    }
}
```
sensor_thread.cpp
```cpp
#include "messages.h"
#include <stdint.h>

#include <zephyr/kernel.h>
#include <zephyr/logging/log.h>
#include <zephyr/zbus/zbus.h>

#define GET_TIME_IN_NS() (k_uptime_get () * 1000)

ZBUS_SUBSCRIBER_DEFINE (sensor_sampler_thread_sub, 4);

ZBUS_CHAN_DEFINE (sensor_data_chan, struct sensor_data_msg, NULL,  NULL, ZBUS_OBSERVERS_EMPTY, ZBUS_MSG_INIT (0));
ZBUS_CHAN_DECLARE (chan_wifi_connection);

void main_thread (){
  const struct zbus_channel *chan;
  struct sensor_data_msg sdata = { .wetness = 0, .temperature = 0 };
  bool wifi_connected = false;
  while (1) {
      zbus_sub_wait (&sensor_sampler_thread_sub, &chan, K_FOREVER);
      zbus_chan_claim (&chan_wifi_connection, K_MSEC (500));
      zbus_chan_read (&chan_wifi_connection, &wifi_connected, K_MSEC (500));
      zbus_chan_finish (&chan_wifi_connection);

      if (!wifi_connected){
          continue;
        }
      sdata.wetness += 1;
      sdata.temperature += 2;
      uint64_t start = GET_TIME_IN_NS ();
      zbus_chan_pub (&sensor_data_chan, &sdata, K_MSEC (500));
      uint64_t delta = GET_TIME_IN_NS () - start;
    }
}
```
listener.cpp
```
#include <zephyr/kernel.h>
#include <zephyr/logging/log.h>
#include <zephyr/zbus/zbus.h>
LOG_MODULE_DECLARE (zbus, CONFIG_ZBUS_LOG_LEVEL);

void listener_cb (const struct zbus_channel *chan)
{
  LOG_INF ("Data received on channel %s", zbus_chan_name (chan));
  LOG_HEXDUMP_INF (zbus_chan_const_msg (chan), zbus_chan_msg_size (chan), "Msg content:");
}

ZBUS_LISTENER_DEFINE (dev_log_lis, listener_cb);
```

**To Reproduce**
Steps to reproduce the behavior:
1. Build the above code in a project
2. Read error log

**Expected behavior**
This code is supposed to simply pass data using Zbus. Initally the wifi is supposed to connect to a given network and having connected, the connection status is sent via `chan_wifi_connection` and this triggers the sensor thread to publish a reading to `sensor_data_chan` when a notification is received from `sample_trigger_chan`. Finally, when data is received is received by the temp_ctrl_thread.cpp, it is printed out to the console. All the while, the listener is following all the messages in `sensor_data_chan` and `chan_wifi_connection`.

**Logs and console output**
**Command**: west build --board esp32s3_devkitm zbus_cpp/ --pristine always
```
-- west build: making build dir /workdir/app/build pristine
-- west build: generating a build system
Loading Zephyr default modules (Zephyr base).
-- Application: /workdir/app/zbus_cpp
-- CMake version: 3.22.1
-- Found Python3: /usr/bin/python3 (found suitable version "3.10.12", minimum required is "3.8") found components: Interpreter 
-- Cache files will be written to: /home/user/.cache/zephyr
-- Zephyr version: 3.5.0 (/workdir/deps/zephyr)
-- Found west (found suitable version "1.0.0", minimum required is "0.14.0")
-- Board: esp32s3_devkitm
-- Found host-tools: zephyr 0.16.1 (/opt/toolchains/zephyr-sdk-0.16.1)
-- Found toolchain: zephyr 0.16.1 (/opt/toolchains/zephyr-sdk-0.16.1)
-- Found Dtc: /opt/toolchains/zephyr-sdk-0.16.1/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.6.0", minimum required is "1.4.6") 
-- Found BOARD.dts: /workdir/deps/zephyr/boards/xtensa/esp32s3_devkitm/esp32s3_devkitm.dts
-- Found devicetree overlay: /workdir/app/zbus_cpp/boards/esp32s3_devkitm.overlay
-- Generated zephyr.dts: /workdir/app/build/zephyr/zephyr.dts
-- Generated devicetree_generated.h: /workdir/app/build/zephyr/include/generated/devicetree_generated.h
-- Including generated dts.cmake file: /workdir/app/build/zephyr/dts.cmake
Parsing /workdir/app/zbus_cpp/Kconfig
Loaded configuration '/workdir/deps/zephyr/boards/xtensa/esp32s3_devkitm/esp32s3_devkitm_defconfig'
Merged configuration '/workdir/app/zbus_cpp/prj.conf'
Merged configuration '/workdir/app/zbus_cpp/boards/esp32s3_devkitm.conf'
Configuration saved to '/workdir/app/build/zephyr/.config'
Kconfig header saved to '/workdir/app/build/zephyr/include/generated/autoconf.h'
-- Found GnuLd: /opt/toolchains/zephyr-sdk-0.16.1/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd (found version "2.38") 
-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- The ASM compiler identification is GNU
-- Found assembler: /opt/toolchains/zephyr-sdk-0.16.1/xtensa-espressif_esp32s3_zephyr-elf/bin/xtensa-espressif_esp32s3_zephyr-elf-gcc
CMake Warning at /workdir/deps/zephyr/CMakeLists.txt:1962 (message):
  __ASSERT() statements are globally ENABLED


-- Configuring done
-- Generating done
-- Build files have been written to: /workdir/app/build
-- west build: building application
[1/361] Preparing syscall dependency handling

[3/361] Generating include/generated/version.h
-- Zephyr version: 3.5.0 (/workdir/deps/zephyr), build: zephyr-v3.5.0
[345/361] Performing configure step for 'EspIdfBootloader'
-- Found Git: /usr/bin/git (found version "2.34.1") 
-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- The ASM compiler identification is GNU
-- Found assembler: /opt/toolchains/zephyr-sdk-0.16.1/xtensa-espressif_esp32s3_zephyr-elf/bin/xtensa-espressif_esp32s3_zephyr-elf-gcc
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /opt/toolchains/zephyr-sdk-0.16.1/xtensa-espressif_esp32s3_zephyr-elf/bin/xtensa-espressif_esp32s3_zephyr-elf-gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /opt/toolchains/zephyr-sdk-0.16.1/xtensa-espressif_esp32s3_zephyr-elf/bin/xtensa-espressif_esp32s3_zephyr-elf-g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Building ESP-IDF components for target esp32s3
-- Project sdkconfig file /workdir/app/build/esp-idf/build/sdkconfig
-- Adding linker script /workdir/deps/modules/hal/espressif/components/soc/esp32s3/ld/esp32s3.peripherals.ld
-- Adding linker script /workdir/deps/modules/hal/espressif/components/esp_rom/esp32s3/ld/esp32s3.rom.ld
-- Adding linker script /workdir/deps/modules/hal/espressif/components/esp_rom/esp32s3/ld/esp32s3.rom.api.ld
-- Adding linker script /workdir/deps/modules/hal/espressif/components/esp_rom/esp32s3/ld/esp32s3.rom.libgcc.ld
-- Adding linker script /workdir/deps/modules/hal/espressif/components/esp_rom/esp32s3/ld/esp32s3.rom.newlib.ld
-- Adding linker script /workdir/deps/modules/hal/espressif/components/bootloader/subproject/main/ld/esp32s3/bootloader.ld
-- Adding linker script /workdir/deps/modules/hal/espressif/components/bootloader/subproject/main/ld/esp32s3/bootloader.rom.ld
-- Components: bootloader bootloader_support efuse esp32s3 esp_common esp_hw_support esp_rom esp_system esptool_py freertos hal log main micro-ecc newlib partition_table soc spi_flash xtensa
-- Component paths: /workdir/deps/modules/hal/espressif/components/bootloader /workdir/deps/modules/hal/espressif/components/bootloader_support /workdir/deps/modules/hal/espressif/components/efuse /workdir/deps/modules/hal/espressif/components/esp32s3 /workdir/deps/modules/hal/espressif/components/esp_common /workdir/deps/modules/hal/espressif/components/esp_hw_support /workdir/deps/modules/hal/espressif/components/esp_rom /workdir/deps/modules/hal/espressif/components/esp_system /workdir/deps/modules/hal/espressif/components/esptool_py /workdir/deps/modules/hal/espressif/components/freertos /workdir/deps/modules/hal/espressif/components/hal /workdir/deps/modules/hal/espressif/components/log /workdir/deps/modules/hal/espressif/components/bootloader/subproject/main /workdir/deps/modules/hal/espressif/components/bootloader/subproject/components/micro-ecc /workdir/deps/modules/hal/espressif/components/newlib /workdir/deps/modules/hal/espressif/components/partition_table /workdir/deps/modules/hal/espressif/components/soc /workdir/deps/modules/hal/espressif/components/spi_flash /workdir/deps/modules/hal/espressif/components/xtensa
-- Configuring done
-- Generating done
-- Build files have been written to: /workdir/app/build/esp-idf/build/bootloader
[346/361] Performing build step for 'EspIdfBootloader'
[1/101] Generating project_elf_src_esp32s3.c
[2/101] Building C object esp-idf/soc/CMakeFiles/__idf_soc.dir/soc_include_legacy_warn.c.obj
[3/101] Building C object esp-idf/soc/CMakeFiles/__idf_soc.dir/esp32s3/gdma_periph.c.obj
[4/101] Building C object esp-idf/soc/CMakeFiles/__idf_soc.dir/esp32s3/dedic_gpio_periph.c.obj
[5/101] Building C object CMakeFiles/bootloader.elf.dir/project_elf_src_esp32s3.c.obj
[6/101] Building C object esp-idf/hal/CMakeFiles/__idf_hal.dir/mpu_hal.c.obj
[7/101] Building C object esp-idf/soc/CMakeFiles/__idf_soc.dir/lldesc.c.obj
[8/101] Building C object esp-idf/soc/CMakeFiles/__idf_soc.dir/esp32s3/adc_periph.c.obj
[9/101] Building C object esp-idf/soc/CMakeFiles/__idf_soc.dir/esp32s3/interrupts.c.obj
[10/101] Building C object esp-idf/soc/CMakeFiles/__idf_soc.dir/esp32s3/lcd_periph.c.obj
[11/101] Building C object esp-idf/soc/CMakeFiles/__idf_soc.dir/esp32s3/i2s_periph.c.obj
[12/101] Building C object esp-idf/soc/CMakeFiles/__idf_soc.dir/esp32s3/gpio_periph.c.obj
[13/101] Building C object esp-idf/hal/CMakeFiles/__idf_hal.dir/cpu_hal.c.obj
[14/101] Building C object esp-idf/soc/CMakeFiles/__idf_soc.dir/esp32s3/i2c_periph.c.obj
[15/101] Building C object esp-idf/soc/CMakeFiles/__idf_soc.dir/esp32s3/rmt_periph.c.obj
[16/101] Building C object esp-idf/soc/CMakeFiles/__idf_soc.dir/esp32s3/pcnt_periph.c.obj
[17/101] Building C object esp-idf/soc/CMakeFiles/__idf_soc.dir/esp32s3/ledc_periph.c.obj
[18/101] Building C object esp-idf/soc/CMakeFiles/__idf_soc.dir/esp32s3/sdmmc_periph.c.obj
[19/101] Building C object esp-idf/soc/CMakeFiles/__idf_soc.dir/esp32s3/mcpwm_periph.c.obj
[20/101] Building C object esp-idf/soc/CMakeFiles/__idf_soc.dir/esp32s3/sigmadelta_periph.c.obj
[21/101] Building C object esp-idf/soc/CMakeFiles/__idf_soc.dir/esp32s3/rtc_io_periph.c.obj
[22/101] Building C object esp-idf/soc/CMakeFiles/__idf_soc.dir/esp32s3/sdio_slave_periph.c.obj
[23/101] Building C object esp-idf/soc/CMakeFiles/__idf_soc.dir/esp32s3/uart_periph.c.obj
[24/101] Building C object esp-idf/spi_flash/CMakeFiles/__idf_spi_flash.dir/esp32s3/spi_flash_rom_patch.c.obj
[25/101] Building C object esp-idf/soc/CMakeFiles/__idf_soc.dir/esp32s3/usb_phy_periph.c.obj
[26/101] Building C object esp-idf/soc/CMakeFiles/__idf_soc.dir/esp32s3/timer_periph.c.obj
[27/101] Building C object esp-idf/soc/CMakeFiles/__idf_soc.dir/esp32s3/spi_periph.c.obj
[28/101] Building C object esp-idf/soc/CMakeFiles/__idf_soc.dir/esp32s3/touch_sensor_periph.c.obj
[29/101] Building C object esp-idf/hal/CMakeFiles/__idf_hal.dir/wdt_hal_iram.c.obj
[30/101] Building C object esp-idf/soc/CMakeFiles/__idf_soc.dir/esp32s3/usb_periph.c.obj
[31/101] Building C object esp-idf/bootloader_support/CMakeFiles/__idf_bootloader_support.dir/src/bootloader_clock_init.c.obj
[32/101] Building C object esp-idf/bootloader_support/CMakeFiles/__idf_bootloader_support.dir/src/bootloader_mem.c.obj
[33/101] Building C object esp-idf/bootloader_support/CMakeFiles/__idf_bootloader_support.dir/src/bootloader_random.c.obj
[34/101] Building C object esp-idf/bootloader_support/CMakeFiles/__idf_bootloader_support.dir/src/bootloader_random_esp32s3.c.obj
[35/101] Building C object esp-idf/bootloader_support/CMakeFiles/__idf_bootloader_support.dir/src/bootloader_common_loader.c.obj
[36/101] Building C object esp-idf/bootloader_support/CMakeFiles/__idf_bootloader_support.dir/src/secure_boot.c.obj
[37/101] Building C object esp-idf/bootloader_support/CMakeFiles/__idf_bootloader_support.dir/src/flash_encrypt.c.obj
[38/101] Building C object esp-idf/bootloader_support/CMakeFiles/__idf_bootloader_support.dir/src/bootloader_common.c.obj
[39/101] Building C object esp-idf/bootloader_support/CMakeFiles/__idf_bootloader_support.dir/src/bootloader_efuse_esp32s3.c.obj
[40/101] Building C object esp-idf/bootloader_support/CMakeFiles/__idf_bootloader_support.dir/src/flash_partitions.c.obj
[41/101] Building C object esp-idf/bootloader_support/CMakeFiles/__idf_bootloader_support.dir/src/bootloader_clock_loader.c.obj
[42/101] Building C object esp-idf/bootloader_support/CMakeFiles/__idf_bootloader_support.dir/src/bootloader_flash_config_esp32s3.c.obj
[43/101] Building C object esp-idf/bootloader_support/CMakeFiles/__idf_bootloader_support.dir/src/flash_qio_mode.c.obj
[44/101] Building C object esp-idf/bootloader_support/CMakeFiles/__idf_bootloader_support.dir/src/bootloader_console_loader.c.obj
[45/101] Building C object esp-idf/bootloader_support/CMakeFiles/__idf_bootloader_support.dir/src/bootloader_console.c.obj
[46/101] Building C object esp-idf/bootloader_support/CMakeFiles/__idf_bootloader_support.dir/src/esp32s3/bootloader_sha.c.obj
[47/101] Building C object esp-idf/bootloader_support/CMakeFiles/__idf_bootloader_support.dir/src/esp32s3/bootloader_soc.c.obj
[48/101] Building C object esp-idf/bootloader_support/CMakeFiles/__idf_bootloader_support.dir/src/bootloader_init.c.obj
[49/101] Building C object esp-idf/bootloader_support/CMakeFiles/__idf_bootloader_support.dir/src/bootloader_flash.c.obj
[50/101] Building C object esp-idf/bootloader_support/CMakeFiles/__idf_bootloader_support.dir/src/bootloader_panic.c.obj
[51/101] Building C object esp-idf/bootloader_support/CMakeFiles/__idf_bootloader_support.dir/src/bootloader_utility.c.obj
[52/101] Building C object esp-idf/efuse/CMakeFiles/__idf_efuse.dir/esp32s3/esp_efuse_fields.c.obj
[53/101] Building C object esp-idf/efuse/CMakeFiles/__idf_efuse.dir/esp32s3/esp_efuse_table.c.obj
[54/101] Building C object esp-idf/efuse/CMakeFiles/__idf_efuse.dir/esp32s3/esp_efuse_rtc_calib.c.obj
[55/101] Building C object esp-idf/bootloader_support/CMakeFiles/__idf_bootloader_support.dir/src/esp32s3/bootloader_esp32s3.c.obj
[56/101] Building C object esp-idf/efuse/CMakeFiles/__idf_efuse.dir/esp32s3/esp_efuse_utility.c.obj
[57/101] Building C object esp-idf/efuse/CMakeFiles/__idf_efuse.dir/src/esp_efuse_fields.c.obj
[58/101] Building C object esp-idf/esp_system/CMakeFiles/__idf_esp_system.dir/esp_err.c.obj
[59/101] Building C object esp-idf/esp_hw_support/CMakeFiles/__idf_esp_hw_support.dir/compare_set.c.obj
[60/101] Building C object esp-idf/bootloader_support/CMakeFiles/__idf_bootloader_support.dir/src/esp_image_format.c.obj
[61/101] Building C object esp-idf/efuse/CMakeFiles/__idf_efuse.dir/src/esp_efuse_api.c.obj
[62/101] Building C object esp-idf/esp_hw_support/CMakeFiles/__idf_esp_hw_support.dir/cpu_util.c.obj
[63/101] Building C object esp-idf/esp_hw_support/CMakeFiles/__idf_esp_hw_support.dir/port/esp32s3/rtc_clk_init.c.obj
[64/101] Building C object esp-idf/esp_hw_support/CMakeFiles/__idf_esp_hw_support.dir/port/esp32s3/rtc_pm.c.obj
[65/101] Building C object esp-idf/esp_hw_support/CMakeFiles/__idf_esp_hw_support.dir/port/esp32s3/rtc_wdt.c.obj
[66/101] Building C object esp-idf/efuse/CMakeFiles/__idf_efuse.dir/src/esp_efuse_utility.c.obj
[67/101] Building C object esp-idf/esp_hw_support/CMakeFiles/__idf_esp_hw_support.dir/port/esp32s3/chip_info.c.obj
[68/101] Building C object esp-idf/xtensa/CMakeFiles/__idf_xtensa.dir/eri.c.obj
[69/101] Building C object esp-idf/esp_hw_support/CMakeFiles/__idf_esp_hw_support.dir/port/esp32s3/rtc_time.c.obj
[70/101] Building C object esp-idf/esp_rom/CMakeFiles/__idf_esp_rom.dir/patches/esp_rom_crc.c.obj
[71/101] Building C object esp-idf/xtensa/CMakeFiles/__idf_xtensa.dir/xt_trax.c.obj
[72/101] Building C object esp-idf/esp_rom/CMakeFiles/__idf_esp_rom.dir/patches/esp_rom_sys.c.obj
[73/101] Building C object esp-idf/esp_hw_support/CMakeFiles/__idf_esp_hw_support.dir/port/esp32s3/rtc_clk.c.obj
[74/101] Building C object esp-idf/esp_rom/CMakeFiles/__idf_esp_rom.dir/patches/esp_rom_tjpgd.c.obj
[75/101] Building C object esp-idf/efuse/CMakeFiles/__idf_efuse.dir/src/esp_efuse_api_key_esp32xx.c.obj
[76/101] Building ASM object esp-idf/esp_rom/CMakeFiles/__idf_esp_rom.dir/patches/esp_rom_longjmp.S.obj
[77/101] Building C object esp-idf/esp_common/CMakeFiles/__idf_esp_common.dir/src/esp_err_to_name.c.obj
[78/101] Building C object esp-idf/esp_rom/CMakeFiles/__idf_esp_rom.dir/patches/esp_rom_uart.c.obj
[79/101] Building C object esp-idf/esp_hw_support/CMakeFiles/__idf_esp_hw_support.dir/port/esp32s3/rtc_sleep.c.obj
[80/101] Building C object esp-idf/esp_hw_support/CMakeFiles/__idf_esp_hw_support.dir/port/esp32s3/rtc_init.c.obj
[81/101] Building C object esp-idf/log/CMakeFiles/__idf_log.dir/log_noos.c.obj
[82/101] Building C object esp-idf/main/CMakeFiles/__idf_main.dir/bootloader_start.c.obj
[83/101] Building C object esp-idf/micro-ecc/CMakeFiles/__idf_micro-ecc.dir/uECC_verify_antifault.c.obj
[84/101] Building C object esp-idf/log/CMakeFiles/__idf_log.dir/log.c.obj
[85/101] Building C object esp-idf/log/CMakeFiles/__idf_log.dir/log_buffers.c.obj
[86/101] Linking C static library esp-idf/log/liblog.a
[87/101] Linking C static library esp-idf/esp_rom/libesp_rom.a
[88/101] Linking C static library esp-idf/esp_common/libesp_common.a
[89/101] Linking C static library esp-idf/xtensa/libxtensa.a
[90/101] Linking C static library esp-idf/esp_hw_support/libesp_hw_support.a
[91/101] Linking C static library esp-idf/esp_system/libesp_system.a
[92/101] Linking C static library esp-idf/efuse/libefuse.a
[93/101] Linking C static library esp-idf/bootloader_support/libbootloader_support.a
[94/101] Linking C static library esp-idf/spi_flash/libspi_flash.a
[95/101] Linking C static library esp-idf/micro-ecc/libmicro-ecc.a
[96/101] Linking C static library esp-idf/soc/libsoc.a
[97/101] Linking C static library esp-idf/hal/libhal.a
[98/101] Linking C static library esp-idf/main/libmain.a
[99/101] Linking C executable bootloader.elf
[100/101] Generating binary image from built executable
esptool.py v4.5
Creating esp32s3 image...
Merged 1 ELF section
Successfully created esp32s3 image.
Generated /workdir/app/build/esp-idf/build/bootloader/bootloader.bin
[101/101] cd /workdir/app/build/esp-idf/build/bootloader/esp-idf/esptool_py && /usr/bin/python3 /workdir/deps/modules/hal/espressif/components/partition_table/check_sizes.py --offset 0x8000 bootloader 0x0 /workdir/app/build/esp-idf/build/bootloader/bootloader.bin
Bootloader binary size 0x53d0 bytes. 0x2c30 bytes (35%) free.
[354/361] Building CXX object CMakeFiles/app.dir/src/wifi_thread.cpp.obj
In file included from /workdir/app/zbus_cpp/src/wifi_thread.cpp:6:
/workdir/app/zbus_cpp/inc/wifi_thread.h:17:12: warning: 'int set_wifi_credentials(const char*, int, const char*, int)' declared 'static' but never defined [-Wunused-function]
   17 | static int set_wifi_credentials (const char *ssid, const int ssid_length,
      |            ^~~~~~~~~~~~~~~~~~~~
/workdir/app/zbus_cpp/src/wifi_thread.cpp:156:1: warning: 'void wifi_disconnect()' defined but not used [-Wunused-function]
  156 | wifi_disconnect (void)
      | ^~~~~~~~~~~~~~~
[356/361] Linking CXX executable zephyr/zephyr_pre0.elf
FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map /workdir/app/build/zephyr/zephyr_pre0.map 
: && ccache /opt/toolchains/zephyr-sdk-0.16.1/xtensa-espressif_esp32s3_zephyr-elf/bin/xtensa-espressif_esp32s3_zephyr-elf-g++  -gdwarf-4 zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.obj -o zephyr/zephyr_pre0.elf  zephyr/CMakeFiles/offsets.dir/./arch/xtensa/core/offsets/offsets.c.obj  -fuse-ld=bfd  -T  zephyr/linker_zephyr_pre0.cmd  -Wl,-Map=/workdir/app/build/zephyr/zephyr_pre0.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/xtensa/core/libarch__xtensa__core.a  zephyr/lib/libc/picolibc/liblib__libc__picolibc.a  zephyr/lib/libc/common/liblib__libc__common.a  zephyr/subsys/random/libsubsys__random.a  zephyr/subsys/net/libsubsys__net.a  zephyr/subsys/net/l2/ethernet/libsubsys__net__l2__ethernet.a  zephyr/subsys/net/l2/wifi/libsubsys__net__l2__wifi.a  zephyr/subsys/net/ip/libsubsys__net__ip.a  zephyr/subsys/net/lib/config/libsubsys__net__lib__config.a  zephyr/subsys/zbus/libsubsys__zbus.a  zephyr/drivers/interrupt_controller/libdrivers__interrupt_controller.a  zephyr/drivers/clock_control/libdrivers__clock_control.a  zephyr/drivers/console/libdrivers__console.a  zephyr/drivers/entropy/libdrivers__entropy.a  zephyr/drivers/gpio/libdrivers__gpio.a  zephyr/drivers/pinctrl/libdrivers__pinctrl.a  zephyr/drivers/serial/libdrivers__serial.a  zephyr/drivers/timer/libdrivers__timer.a  zephyr/drivers/wifi/libdrivers__wifi.a  modules/mbedtls/libmbedTLSBase.a  modules/mbedtls/libmbedTLSCrypto.a  modules/mbedtls/libmbedTLSX509.a  modules/mbedtls/libmodules__mbedtls.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  -L"/opt/toolchains/zephyr-sdk-0.16.1/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0"  -L/workdir/app/build/zephyr  -lgcc  zephyr/arch/common/libisr_tables.a  modules/mbedtls/libmbedTLSBase.a  modules/mbedtls/libmbedTLSCrypto.a  modules/mbedtls/libmbedTLSX509.a  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn  -lstdc++  -Wl,-no-pie  -DPICOLIBC_INTEGER_PRINTF_SCANF  -T/workdir/deps/modules/hal/espressif/zephyr/esp32s3/src/linker/esp32s3.rom.alias.ld  -T/workdir/deps/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.ld  -T/workdir/deps/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.api.ld  -T/workdir/deps/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.libgcc.ld  -T/workdir/deps/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.newlib.ld  -T/workdir/deps/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.version.ld  -T/workdir/deps/modules/hal/espressif/zephyr/esp32s3/../../components/soc/esp32s3/ld/esp32s3.peripherals.ld  -Wl,--wrap=k_calloc  -Wl,--wrap=k_malloc  -lnet80211  -lcore  -lpp  -lcoexist  -lphy  -lgcc  -L/workdir/deps/modules/hal/espressif/zephyr/esp32s3/../blobs/lib/esp32s3  --specs=picolibc.specs  -lc  -lgcc && cd /workdir/app/build/zephyr && /usr/bin/cmake -E true
/opt/toolchains/zephyr-sdk-0.16.1/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: app/libapp.a(main.cpp.obj):(._zbus_channel_observation.static.chan_wifi_connectionzz3dev_log_lis_+0x0): undefined reference to `chan_wifi_connection'
/opt/toolchains/zephyr-sdk-0.16.1/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: app/libapp.a(main.cpp.obj):(._zbus_channel_observation.static.sensor_data_chanzz3dev_log_lis_+0x0): undefined reference to `sensor_data_chan'
/opt/toolchains/zephyr-sdk-0.16.1/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: app/libapp.a(sensor_sampler_thread.cpp.obj):(.literal._Z21sensor_sampler_threadv+0x4): undefined reference to `chan_wifi_connection'
/opt/toolchains/zephyr-sdk-0.16.1/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: app/libapp.a(temp_ctrl_thread.cpp.obj):(.literal._Z16temp_ctrl_threadv+0x4): undefined reference to `sensor_data_chan'
/opt/toolchains/zephyr-sdk-0.16.1/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: app/libapp.a(temp_ctrl_thread.cpp.obj):(._zbus_channel_observation.static.sensor_data_chanzz3temp_ctrl_thread_sub_+0x0): undefined reference to `sensor_data_chan'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

**Environment (please complete the following information):**
 - OS: Linux (Zephyr Docker Image)
 - Zephyr 3.5.0

--- a/include/zephyr/zbus/zbus.h
+++ b/include/zephyr/zbus/zbus.h
@@ -304,6 +304,7 @@ struct zbus_channel_observation {
 	static struct zbus_channel_data _CONCAT(_zbus_chan_data_, _name) = {                       \
 		.observers_start_idx = -1, .observers_end_idx = -1};                               \
 	static K_MUTEX_DEFINE(_CONCAT(_zbus_mutex_, _name));                                       \
+	IF_ENABLED(CONFIG_CPP, (extern))                                                           \
 	const STRUCT_SECTION_ITERABLE(zbus_channel, _name) = {                                     \
 		ZBUS_CHANNEL_NAME_INIT(_name) /* Maybe removed */                                  \
 			.message = &_CONCAT(_zbus_message_, _name),                                \

